File: t/VB/SVGDemo/Utilities/DocumentText.vb
' *************************************************************
' Copyright (c) 1991-2019 LEAD Technologies, Inc.              
' All Rights Reserved.                                         
' *************************************************************

Imports System.Collections.Generic
Imports System.Text

Imports Leadtools
Imports Leadtools.Svg
Imports System

Namespace SvgDemo
   <Flags()> _
   Public Enum DocumentCharacterPositions
      None
      EndOfWord
      EndOfLine
   End Enum

   <Serializable()> _
   Public Structure DocumentCharacter
      Private _code As Char
      Public Property Code() As Char
         Get
            Return _code
         End Get
         Set(value As Char)
            _code = value
         End Set
      End Property

      Private _bounds As LeadRectD
      Public Property Bounds() As LeadRectD
         Get
            Return _bounds
         End Get
         Set(value As LeadRectD)
            _bounds = value
         End Set
      End Property

      Private _positions As DocumentCharacterPositions
      Public Property Positions() As DocumentCharacterPositions
         Get
            Return _positions
         End Get
         Set(value As DocumentCharacterPositions)
            _positions = value
         End Set
      End Property

      Private _direction As SvgCharacterDirection
      Public Property Direction() As SvgCharacterDirection
         Get
            Return _direction
         End Get
         Set(value As SvgCharacterDirection)
            _direction = value
         End Set
      End Property
   End Structure

   <Serializable()> _
   Public Structure DocumentWord
      Private _value As String
      Public Property Value() As String
         Get
            Return _value
         End Get
         Set(value As String)
            _value = value
         End Set
      End Property

      Private _bounds As LeadRectD
      Public Property Bounds() As LeadRectD
         Get
            Return _bounds
         End Get
         Set(value As LeadRectD)
            _bounds = value
         End Set
      End Property

      Private _firstCharacterIndex As Integer
      Public Property FirstCharacterIndex() As Integer
         Get
            Return _firstCharacterIndex
         End Get
         Set(value As Integer)
            _firstCharacterIndex = value
         End Set
      End Property

      Private _lastCharacterIndex As Integer
      Public Property LastCharacterIndex() As Integer
         Get
            Return _lastCharacterIndex
         End Get
         Set(value As Integer)
            _lastCharacterIndex = value
         End Set
      End Property
   End Structure

   Public Class DocumentText
      Private Sub New(characters As List(Of DocumentCharacter))
         _characters = characters
      End Sub

      Public Sub New()
         _characters = New List(Of DocumentCharacter)()
      End Sub

      Public Shared Function FromSvgDocument(document As SvgDocument) As DocumentText
         ' Out list of charcaters
         Dim characters As List(Of DocumentCharacter) = New List(Of DocumentCharacter)()

         ' To strip whitespaces
         Dim hasWhitespaces As Boolean = False
         Dim prevCharacter As DocumentCharacter = New DocumentCharacter()

         ' Callback
         Dim sortCallback As SvgSortElementsCallback = Function(callabackDocument, info, userData)
                                                          ' Is it text?
                                                          Dim textData As SvgTextData = info.TextData
                                                          If textData IsNot Nothing Then
                                                             ' yes, does it have any characters?
                                                             Dim length As Integer = textData.Text.Length
                                                             For i As Integer = 0 To length - 1
                                                                ' Yes, convert to DocumentCharacter and add them to the list
                                                                Dim flags As SvgTextCharacterFlags = textData.CharacterFlags(i)
                                                                Dim positions As DocumentCharacterPositions = DocumentCharacterPositions.None
                                                                If (flags And SvgTextCharacterFlags.EndOfWord) = SvgTextCharacterFlags.EndOfWord Then
                                                                   positions = positions Or DocumentCharacterPositions.EndOfWord
                                                                End If
                                                                If (flags And SvgTextCharacterFlags.EndOfLine) = SvgTextCharacterFlags.EndOfLine Then
                                                                   positions = positions Or DocumentCharacterPositions.EndOfLine
                                                                End If

                                                                Dim character As DocumentCharacter = New DocumentCharacter() With { _
                                                                   .Code = textData.Text(i), _
                                                                   .Bounds = textData.Bounds(i), _
                                                                   .Positions = positions, _
                                                                   .Direction = textData.Directions(i) _
                                                                }

                                                                ' If the character duplicated, then remove the old one
                                                                If prevCharacter.Bounds = character.Bounds AndAlso prevCharacter.Code = character.Code AndAlso prevCharacter.Direction = character.Direction Then
                                                                   characters.Remove(prevCharacter)
                                                                End If

                                                                characters.Add(character)

                                                                If Not hasWhitespaces AndAlso [Char].IsWhiteSpace(character.Code) Then
                                                                   hasWhitespaces = True
                                                                End If

                                                                prevCharacter = character
                                                             Next
                                                          End If

                                                          Return True

                                                       End Function

         ' Accept text elements only.
         Dim filterElementsCallback As SvgFilterElementsCallback = Function(svgDocument, node, userData)
                                                                      ' Accept text elements only.
                                                                      If node.ElementType <> SvgElementType.Text Then
                                                                         Return False
                                                                      End If

                                                                      Return True

                                                                   End Function

         ' Setup options and sort
         Dim options As SvgSortOptions = New SvgSortOptions() With { _
            .ExtractText = SvgExtractText.Character, _
            .SortFlags = SvgSortFlags.[Default] _
         }

         document.SetFilterElementsCallback(filterElementsCallback, Nothing)
         document.SortElements(options, sortCallback, Nothing)
         document.SetFilterElementsCallback(Nothing, Nothing)

         ' Characters is set, before using it, strip the whitespaces
         If hasWhitespaces Then
            Dim temp As List(Of DocumentCharacter) = New List(Of DocumentCharacter)()
            Dim count As Integer = characters.Count
            Dim positions As DocumentCharacterPositions = DocumentCharacterPositions.None
            Dim newIndex As Integer = -1
            For index As Integer = 0 To count - 1
               Dim character As DocumentCharacter = characters(index)

               ' If this character is a whitespace, dont add it, just steal its position and add it to the previous character
               If [Char].IsWhiteSpace(character.Code) Then
                  positions = positions Or character.Positions
               Else
                  ' If have stolen flags from dropping white space characters, add them to the last non-white space character
                  If positions <> DocumentCharacterPositions.None Then
                     If newIndex <> -1 Then
                        Dim tempCharacter As DocumentCharacter = temp(newIndex)
                        tempCharacter.Positions = tempCharacter.Positions Or positions
                        temp(newIndex) = tempCharacter
                     End If

                     positions = DocumentCharacterPositions.None
                  End If

                  ' Add it
                  temp.Add(character)
                  newIndex += 1
               End If
            Next

            characters = temp
         End If

         Dim result As DocumentText = New DocumentText(characters)
         Return result
      End Function

      Private _characters As List(Of DocumentCharacter)

      Public ReadOnly Property Characters() As IList(Of DocumentCharacter)
         Get
            Return _characters
         End Get
      End Property

      Private _words As IList(Of DocumentWord)

      Public ReadOnly Property Words() As IList(Of DocumentWord)
         Get
            Return _words
         End Get
      End Property

      Public Sub BuildWords()
         Dim characterIndex As Integer = 0
         Dim characterCount As Integer = _characters.Count

         Dim words As List(Of DocumentWord) = New List(Of DocumentWord)()
         Dim lastReverseWord As Boolean = False
         Dim startReverseIndex As Integer = -1

         ' Loop through all the characters
         While characterIndex < characterCount
            ' Find the total bounding rectangle, begin and end index of the next word
            Dim wordBounds As LeadRectD = LeadRectD.Empty
            Dim beginCharacterIndex As Integer = characterIndex
            Dim reverseWord As Boolean = False

            ' Loop till we reach the EndOfWord of EndOfLine
            Dim more As Boolean = True
            Dim character As DocumentCharacter = New DocumentCharacter()
            While more AndAlso characterIndex < characterCount
               character = _characters(characterIndex)

               ' Add the bounding rectangle of this character
               Dim characterBounds As LeadRectD = character.Bounds

               If wordBounds.IsEmpty Then
                  wordBounds = characterBounds
               Else
                  wordBounds = LeadRectD.UnionRects(wordBounds, characterBounds)
               End If

               more = characterIndex < characterCount AndAlso (character.Positions And DocumentCharacterPositions.EndOfWord) <> DocumentCharacterPositions.EndOfWord AndAlso (character.Positions And DocumentCharacterPositions.EndOfLine) <> DocumentCharacterPositions.EndOfLine

               characterIndex += 1

               If character.Direction = SvgCharacterDirection.RightToLeft Then
                  reverseWord = True
               End If
            End While

            ' From the begin and end index, collect the characters into a string
            Dim sb As StringBuilder = New StringBuilder()
            For i As Integer = beginCharacterIndex To characterIndex - 1
               sb.Append(_characters(i).Code)
            Next

            Dim word As DocumentWord = New DocumentWord()
            word.Value = sb.ToString()
            word.Bounds = wordBounds
            word.FirstCharacterIndex = beginCharacterIndex
            word.LastCharacterIndex = characterIndex - 1

            If reverseWord Then
               Dim charArray As Char() = word.Value.ToCharArray()
               Dim startIndex As Integer = 0
               For i As Integer = 0 To charArray.Length - 1
                  Dim nextIndex As Integer = i + 1
                  If _characters(i + word.FirstCharacterIndex).Direction = SvgCharacterDirection.RightToLeft AndAlso (nextIndex >= charArray.Length OrElse _characters(nextIndex + word.FirstCharacterIndex).Direction = SvgCharacterDirection.LeftToRight) Then
                     Array.Reverse(charArray, startIndex, nextIndex - startIndex)
                     startIndex = i
                  End If
               Next

               word.Value = New String(charArray)
            End If


            If lastReverseWord AndAlso reverseWord Then
               startReverseIndex = If(startReverseIndex = -1, words.Count - 1, startReverseIndex)
               words.Insert(startReverseIndex, word)
            Else
               If startReverseIndex <> -1 Then
                  ReverseCharacterPositions(words(startReverseIndex).LastCharacterIndex, words(words.Count - 1).LastCharacterIndex)
                  startReverseIndex = -1
               End If
               words.Add(word)
            End If

            lastReverseWord = reverseWord AndAlso (character.Positions And DocumentCharacterPositions.EndOfLine) <> DocumentCharacterPositions.EndOfLine
         End While

         If startReverseIndex <> -1 Then
            ReverseCharacterPositions(words(startReverseIndex).LastCharacterIndex, words(words.Count - 1).LastCharacterIndex)
         End If

         _words = words
      End Sub

      Private Sub ReverseCharacterPositions(index1 As Integer, index2 As Integer)
         Dim tempPos As DocumentCharacterPositions = _characters(index1).Positions

         Dim tempCharacter As DocumentCharacter = _characters(index1)
         tempCharacter.Positions = _characters(index2).Positions
         _characters(index1) = tempCharacter

         tempCharacter = _characters(index2)
         tempCharacter.Positions = tempPos
         _characters(index2) = tempCharacter
      End Sub

      Public Function BuildText() As String
         If _words IsNot Nothing AndAlso _characters Is Nothing Then
            BuildWords()
         End If

         If _words Is Nothing OrElse _words.Count = 0 Then
            Return String.Empty
         End If

         Dim sb As StringBuilder = New StringBuilder()
         Dim wordCount As Integer = _words.Count
         Dim wordIndex As Integer = 0
         While wordIndex < wordCount
            Dim word As DocumentWord = _words(wordIndex)
            If Not String.IsNullOrEmpty(word.Value) Then
               sb.Append(word.Value)
            End If

            Dim positions As Integer = DocumentCharacterPositions.None
            If word.LastCharacterIndex <> -1 Then
               positions = _characters(word.LastCharacterIndex).Positions
            End If

            Dim lastWord As Boolean = (wordIndex = wordCount - 1)

            If (positions And DocumentCharacterPositions.EndOfLine) = DocumentCharacterPositions.EndOfLine Then
               If Not lastWord Then
                  sb.AppendLine()
               End If
            ElseIf Not lastWord Then
               sb.Append(" ")
            End If

            wordIndex += 1
         End While

         Return sb.ToString()
      End Function
   End Class
End Namespace

File: t/CS/SVGDemo/Utilities/DocumentText.cs
// *************************************************************
// Copyright (c) 1991-2019 LEAD Technologies, Inc.              
// All Rights Reserved.                                         
// *************************************************************
using System;
using System.Collections.Generic;
using System.Text;

using Leadtools;
using Leadtools.Svg;

namespace SvgDemo
{
   [Flags]
   public enum DocumentCharacterPositions
   {
      None,
      EndOfWord,
      EndOfLine
   }

   [Serializable]
   public struct DocumentCharacter
   {
      private char _code;
      public char Code
      {
         get { return _code; }
         set { _code = value; }
      }

      private LeadRectD _bounds;
      public LeadRectD Bounds
      {
         get { return _bounds; }
         set { _bounds = value; }
      }

      private DocumentCharacterPositions _positions;
      public DocumentCharacterPositions Positions
      {
         get { return _positions; }
         set { _positions = value; }
      }

      private SvgCharacterDirection _direction;
      public SvgCharacterDirection Direction
      {
         get { return _direction; }
         set { _direction = value; }
      }
   }

   [Serializable]
   public struct DocumentWord
   {
      private string _value;
      public string Value
      {
         get { return _value; }
         set { _value = value; }
      }

      private LeadRectD _bounds;
      public LeadRectD Bounds
      {
         get { return _bounds; }
         set { _bounds = value; }
      }

      private int _firstCharacterIndex;
      public int FirstCharacterIndex
      {
         get { return _firstCharacterIndex; }
         set { _firstCharacterIndex = value; }
      }

      private int _lastCharacterIndex;
      public int LastCharacterIndex
      {
         get { return _lastCharacterIndex; }
         set { _lastCharacterIndex = value; }
      }
   }

   public class DocumentText
   {
      private DocumentText(List<DocumentCharacter> characters)
      {
         _characters = characters;
      }

      public DocumentText()
      {
         _characters = new List<DocumentCharacter>();
      }

      public static DocumentText FromSvgDocument(SvgDocument document)
      {
         // Out list of charcaters
         var characters = new List<DocumentCharacter>();

         // To strip whitespaces
         var hasWhitespaces = false;
         var prevCharacter = new DocumentCharacter();

         // Callback
         SvgSortElementsCallback sortCallback = (callabackDocument, info, userData) =>
         {
            // Is it text?
            var textData = info.TextData;
            if (textData != null)
            {
               // yes, does it have any characters?
               var length = textData.Text.Length;
               for (var i = 0; i < length; i++)
               {
                  // Yes, convert to DocumentCharacter and add them to the list
                  var flags = textData.CharacterFlags[i];
                  var positions = DocumentCharacterPositions.None;
                  if ((flags & SvgTextCharacterFlags.EndOfWord) == SvgTextCharacterFlags.EndOfWord)
                     positions |= DocumentCharacterPositions.EndOfWord;
                  if ((flags & SvgTextCharacterFlags.EndOfLine) == SvgTextCharacterFlags.EndOfLine)
                     positions |= DocumentCharacterPositions.EndOfLine;

                  var character = new DocumentCharacter
                  {
                     Code = textData.Text[i],
                     Bounds = textData.Bounds[i],
                     Positions = positions,
                     Direction = textData.Directions[i]
                  };

                  // If the character duplicated, then remove the old one
                  if (prevCharacter.Bounds == character.Bounds &&
                     prevCharacter.Code == character.Code &&
                     prevCharacter.Direction == character.Direction)
                  {
                     characters.Remove(prevCharacter);
                  }

                  characters.Add(character);

                  if (!hasWhitespaces && Char.IsWhiteSpace(character.Code))
                     hasWhitespaces = true;

                  prevCharacter = character;
               }
            }

            return true;
         };

         // Accept text elements only.
         SvgFilterElementsCallback filterElementsCallback = (svgDocument, nodeHandle, userData) =>
         {
            // Accept text elements only.
            if (nodeHandle.ElementType != SvgElementType.Text)
               return false;

            return true;
         };

         // Setup options and sort
         var options = new SvgSortOptions { ExtractText = SvgExtractText.Character, SortFlags = SvgSortFlags.Default };
         document.SetFilterElementsCallback(filterElementsCallback, null);
         document.SortElements(options, sortCallback, null);
         document.SetFilterElementsCallback(null, null);

         // Characters is set, before using it, strip the whitespaces
         if (hasWhitespaces)
         {
            var temp = new List<DocumentCharacter>();
            var count = characters.Count;
            var positions = DocumentCharacterPositions.None;
            var newIndex = -1;
            for (var index = 0; index < count; index++)
            {
               var character = characters[index];

               // If this character is a whitespace, dont add it, just steal its position and add it to the previous character
               if (Char.IsWhiteSpace(character.Code))
               {
                  positions |= character.Positions;
               }
               else
               {
                  // If have stolen flags from dropping white space characters, add them to the last non-white space character
                  if (positions != DocumentCharacterPositions.None)
                  {
                     if (newIndex != -1)
                     {
                        var tempCharacter = temp[newIndex];
                        tempCharacter.Positions |= positions;
                        temp[newIndex] = tempCharacter;
                     }

                     positions = DocumentCharacterPositions.None;
                  }

                  // Add it
                  temp.Add(character);
                  newIndex++;
               }
            }

            characters = temp;
         }

         var result = new DocumentText(characters);
         return result;
      }

      private List<DocumentCharacter> _characters;

      public IList<DocumentCharacter> Characters
      {
         get { return _characters; }
      }

      private IList<DocumentWord> _words;

      public IList<DocumentWord> Words
      {
         get { return _words; }
      }

      public void BuildWords()
      {
         var characterIndex = 0;
         var characterCount = _characters.Count;

         var words = new List<DocumentWord>();
         var lastReverseWord = false;
         var startReverseIndex = -1;

         // Loop through all the characters
         while(characterIndex < characterCount)
         {
            // Find the total bounding rectangle, begin and end index of the next word
            var wordBounds = LeadRectD.Empty;
            var beginCharacterIndex = characterIndex;
            bool reverseWord = false;

            // Loop till we reach the EndOfWord of EndOfLine
            var more = true;
            var character = new DocumentCharacter();
            while (more && characterIndex < characterCount)
            {
               character = _characters[characterIndex];

               // Add the bounding rectangle of this character
               var characterBounds = character.Bounds;

               if (wordBounds.IsEmpty)
                  wordBounds = characterBounds;
               else
                  wordBounds = LeadRectD.UnionRects(wordBounds, characterBounds);

               more = characterIndex < characterCount &&
                     (character.Positions & DocumentCharacterPositions.EndOfWord) != DocumentCharacterPositions.EndOfWord &&
                     (character.Positions & DocumentCharacterPositions.EndOfLine) != DocumentCharacterPositions.EndOfLine;

               characterIndex++;

               if (character.Direction == SvgCharacterDirection.RightToLeft)
                  reverseWord = true;
            }

            // From the begin and end index, collect the characters into a string
            var sb = new StringBuilder();
            for(var i = beginCharacterIndex; i < characterIndex; i++)
               sb.Append(_characters[i].Code);

            var word = new DocumentWord();
            word.Value = sb.ToString();
            word.Bounds = wordBounds;
            word.FirstCharacterIndex = beginCharacterIndex;
            word.LastCharacterIndex = characterIndex - 1;

            if (reverseWord)
            {
               char[] charArray = word.Value.ToCharArray();
               int startIndex = 0;
               for (int i = 0; i < charArray.Length; i++)
               {
                  int nextIndex = i + 1;
                  if (_characters[i + word.FirstCharacterIndex].Direction == SvgCharacterDirection.RightToLeft &&
                      (nextIndex >= charArray.Length ||
                      _characters[nextIndex + word.FirstCharacterIndex].Direction == SvgCharacterDirection.LeftToRight))
                  {
                     Array.Reverse(charArray, startIndex, nextIndex - startIndex);
                     startIndex = i;
                  }
               }

               word.Value = new string(charArray);
            }


            if (lastReverseWord && reverseWord)
            {
               startReverseIndex = startReverseIndex == -1 ? words.Count - 1: startReverseIndex;
               words.Insert(startReverseIndex, word);
            }
            else
            {
               if (startReverseIndex != -1)
               {
                  ReverseCharacterPositions(words[startReverseIndex].LastCharacterIndex, words[words.Count - 1].LastCharacterIndex);
                  startReverseIndex = -1;
               }
               words.Add(word);
            }

            lastReverseWord = reverseWord && (character.Positions & DocumentCharacterPositions.EndOfLine) != DocumentCharacterPositions.EndOfLine;
         }

         if (startReverseIndex != -1)
            ReverseCharacterPositions(words[startReverseIndex].LastCharacterIndex, words[words.Count - 1].LastCharacterIndex);

         _words = words;
      }

      private void ReverseCharacterPositions(int index1, int index2)
      {
         DocumentCharacterPositions tempPos = _characters[index1].Positions;

         DocumentCharacter tempCharacter = _characters[index1];
         tempCharacter.Positions = _characters[index2].Positions;
         _characters[index1] = tempCharacter;

         tempCharacter = _characters[index2];
         tempCharacter.Positions = tempPos;
         _characters[index2] = tempCharacter;
      }

      public string BuildText()
      {
         if (_words != null && _characters == null)
            BuildWords();

         if (_words == null || _words.Count == 0)
            return string.Empty;

         var sb = new StringBuilder();
         var wordCount = _words.Count;
         var wordIndex = 0;
         while (wordIndex <wordCount)
         {
            var word = _words[wordIndex];
            if (!string.IsNullOrEmpty(word.Value))
               sb.Append(word.Value);

            var positions = DocumentCharacterPositions.None;
            if (word.LastCharacterIndex != -1)
               positions = _characters[word.LastCharacterIndex].Positions;

            var lastWord = (wordIndex == wordCount - 1);

            if ((positions & DocumentCharacterPositions.EndOfLine) == DocumentCharacterPositions.EndOfLine)
            {
               if (!lastWord)
                  sb.AppendLine();
            }
            else if (!lastWord)
            {
               sb.Append(" ");
            }

            wordIndex++;
         }

         return sb.ToString();
      }
   }
}
